src/lib/routers/history.ts

```diff
@@ -179,7 +179,7 @@ class BrowserHistory {
   /**
    * Removes the event listener and cleans up the URL.
    */
-  public dispose(): void {
+  public dispose(): undefined {
```

Comment:
I pretty sure that you should keep `void` here and that it is the `Router.dispose` signature in `src/types/instantsearch.ts` that is incorrect.

I think this is what you wanted:

```diff
 export type Router<TRouteState = UiState> = {
   onUpdate(callback: (route: TRouteState) => void): void;
   read(): UiState;
   write(route: TRouteState): void;
   createURL(state: TRouteState): string;
-  dispose(): Widget['dispose'];
+  dispose: Widget['dispose'];
 };
```

Then there is multiple differences between `Router.dispose` and `Widget.dispose`: 

* mandatory on `Router` and optional on `Widget`
* `options` parameter on `Widget` and no parameters on `Router`

So the final definition should be more like:

```diff
 export type Router<TRouteState = UiState> = {
   onUpdate(callback: (route: TRouteState) => void): void;
   read(): UiState;
   write(route: TRouteState): void;
   createURL(state: TRouteState): string;
-  dispose(): Widget['dispose'];
+  dispose: SearchParameters | void;
 };
```

WDYT?

Reply:
You're right, this is not optimal. I tried another approach in 611090bff85eb37e8b5f8d99dbc65d2cb3561fa6: it's made more generic and the `Router` extends the `Widget` type. I believe this is more accurate. Let me know what you think.

Reply:
LGTM :+1: